### ISSUE TITLE
Thread stuck and crash with basic auth backend 

### ISSUE THREAD

**Issue by bsundmark** [haproxy/haproxy#1298]
## Detailed description of the problem

We are running haproxy alpine docker container from docker hub as an ECS task in AWS where we have one backend that use basic auth for access control. We have other backends not using authentication and backends where we use certificate based authentication and we have never experienced any problems with that.

We have been running this basic auth service for several months with very low volumes of calls coming in and it has been working fine as far as we know. Volumes has been < 20-30 calls / 24h.

A few days ago a change was implemented in our applications which started to use this backend more frequently with a batch job for a limited time. We experienced that haproxy started crashing with messages of a thread stuck and the basic auth backend was refered in the error messages.

We were running two instances in parallell on separate EC2 instances (VMs) with an AWS Network Load Balancer in front of them and both instances crashed more or less simultaneously.

We then updated haproxy from 2.0.22-alpine to the latest 2.4.0-alpine container but the crashes still occured and we also increased the number of haproxy instances from 2 to 4 as it seemed like the problem was related to the load.

This morning the batch job was run again and all 4 containers crashed almost immediately with similar errors. One of them apparently after receiving only 3 calls to this endpoint. New instances were started and the rest of the batch with a total of apx 1000 calls spread over the 4 instances ran to completion without any further crashes.

## Expected behavior

Haproxy should not crash

## Steps to reproduce the behavior

1. 
2. 
3. 

We have not yet tried to reproduce it outside of our production environment. Will start on that ASAP.

## Do you have any idea what may have caused this?

No.
From the behaviour it seems as if basic auth in haproxy can not cope with calls coming in almost simultaneously. It might also be that it is more likely after a period of inactivity but this is speculation.

## Do you have an idea how to solve the issue?

No.

## What is your configuration?

```
#This file is for haproxy-external in AWS
global
  chroot  /var/lib/haproxy
  log  stdout local2
  log-send-hostname haproxy-external-prod-aws

defaults
  default-server init-addr none inter 5s fastinter 2s downinter 2s
  log  global
  maxconn  2000
  mode  http
  option  httplog
  option  dontlognull
  option  tcpka
  option  redispatch
  option  forwardfor
  option  httpchk GET /admin/healthcheck
  retries  3
  timeout  connect 5s
  timeout  client 600s
  timeout  server 600s
  timeout  check 5s

resolvers NovaDNS
  parse-resolv-conf
  hold valid 10s
  accepted_payload_size 8192 # allow larger DNS payloads

userlist basic-auth-list
  group is-basicauth-user
  user bauth password <HASHED-PWD>

frontend ssl-unsecure-in
  bind 0.0.0.0:4443 accept-proxy ssl crt /usr/local/etc/pki/haproxy_public.hapem ca-file /usr/local/etc/pki/truststore.pem verify none

  capture request header X-LOG-TRANSACTION-ID len 36
  capture request header X-User len 128

  acl is_api_ext                        base_beg -i api-ext.
  acl is_customer_ext                   base_beg -i customer.
  acl is_deny_path                      path_beg -i /doc /health /service-configuration /service-metrics /version
  acl is_bauth_notification             path_beg -i /bauth/notification

  http-request set-header               X-Forwarded-For %[src]
  http-request set-header               X-Real-IP %[src]
  http-request deny                     if is_deny_path

  use_backend bauth-be                  if is_api_ext is_bauth_notification
  use_backend customer-admin-be         if is_customer_ext
  use_backend denied

frontend ssl-secure-in
  bind 0.0.0.0:443 accept-proxy ssl crt /usr/local/etc/pki/haproxy_public.hapem ca-file /usr/local/etc/pki/truststore.pem verify required

  capture request header X-LOG-TRANSACTION-ID len 36
  capture request header X-User len 128

  acl is_api                                base_beg -i api.
  acl is_api_client                         hdr_end(x-user) -i CN=public.net user1@company.com user2@company.com user3@company.com
  acl is_api_extsec                         base_beg -i api-extsec.
  acl is_service_ext                        base_beg -i service-ext.
  acl is_endpoint1_communication            path_beg -i /communication/endpoint1
  acl is_endpoint1_trusted_client           hdr_end(x-user) -i CN=trusted-cert user1@company.com user2@company.com user3@company.com
  acl is_hastats                            path_beg -i /hastats
  acl is_hastats_client                     hdr_end(x-user) -i @company.com CN=monitoring-client

  http-request set-header X-User %{+Q}[ssl_c_s_dn]
  http-request set-header X-Forwarded-For %[src]
  http-request set-header X-Real-IP %[src]

  http-request allow                        if is_api or is_api_extsec or is_service_ext
  http-request allow                        if is_endpoint1_trusted_client or is_hastats_client
  http-request deny

  use_backend service-be                    if is_api is_api_client
  use_backend local_hastats                 if is_hastats is_hastats_client
  use_backend communication-be              if is_api_extsec is_endpoint1_communication
  use_backend denied

listen stats
  bind 0.0.0.0:8001
  mode http
  stats enable
  stats uri /hastats
  stats refresh 5s

listen health
  bind *:80
  monitor-uri   /healthcheck
  http-request  deny

backend bauth-be
  option httpchk GET /bauth/admin/healthcheck
  acl bauth-auth http_auth(basic-auth-list)
  http-request auth realm Bauth unless bauth-auth
  server-template bauth-backend 1-8 _bauth._tcp.sd.company.com check resolvers NovaDNS init-addr none verify none

backend communication-be
  option httpchk GET /communication/admin/healthcheck
  server-template communication-backend 1-8 _communication._tcp.sd.company.com check resolvers NovaDNS init-addr none verify none

backend customer-admin-be
  option httpchk GET /admin/healthcheck HTTP/1.0
  server-template customer-admin-be 1-8 _customeradmin._tcp.sd.company.com check resolvers NovaDNS init-addr none verify none

backend service-be
  option httpchk GET /healthcheck

  server prod-aws-nlb1 service.company.com:443 check resolvers NovaDNS ssl verify required crt /usr/local/etc/pki/haproxy.hapem ca-file /usr/local/etc/pki/truststore.pem
  server prod-aws-nlb2 service.company.com:443 check resolvers NovaDNS ssl verify required crt /usr/local/etc/pki/haproxy.hapem ca-file /usr/local/etc/pki/truststore.pem
  server prod-aws-nlb3 service.company.com:443 check resolvers NovaDNS ssl verify required crt /usr/local/etc/pki/haproxy.hapem ca-file /usr/local/etc/pki/truststore.pem

backend denied
  http-request deny

backend local_hastats
  server local 127.0.0.1:8001/hastats
```

## Output of `haproxy -vv` and `uname -a`

```
# haproxy -vv
HAProxy version 2.4.0-6cbbecf 2021/05/14 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.0.html
Running on: Linux 4.14.232-176.381.amzn2.x86_64 #1 SMP Wed May 19 00:31:54 UTC 2021 x86_64
Build options :
  TARGET  = linux-musl
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=2).
Built with OpenSSL version : OpenSSL 1.1.1k  25 Mar 2021
Running on OpenSSL version : OpenSSL 1.1.1k  25 Mar 2021
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.36 2020-12-04
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.2.1 20201203

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace

# uname -a
Linux da9030b069eb 4.14.232-176.381.amzn2.x86_64 #1 SMP Wed May 19 00:31:54 UTC 2021 x86_64 Linux
```

## If HAProxy crashed: Last outputs and backtraces

```
Order of lines are reversed as this comes from our logserver. AWS ECS deletes the container catalogue with the logfile when the task dies unfortunately.

2021-06-21 00:19:13.668	[ALERT]    (1) : Current worker #1 (25) exited with code 134 (Aborted)
2021-06-21 00:19:13.668	               fct=0x55d1532a6200(process_stream) ctx=0x7f68f346fda0
2021-06-21 00:19:13.668	             cob=0,0:NONE(0)/NONE(0)/NONE(0)
2021-06-21 00:19:13.668	[NOTICE]   (1) : haproxy version is 2.4.0-6cbbecf
2021-06-21 00:19:13.668	             strm=0x7f68f346fda0 src=<SRC-IP> fe=ssl-unsecure-in be=bauth-be dst=unknown
2021-06-21 00:19:13.668	             af=0,0 csf=0x7f68f33cf460,4000
2021-06-21 00:19:13.668	             cof=0x7f68f35836a0,a0003300:H1(0x7f68f35bfc50)/SSL(0x7f68f353f1c0)/tcpv4(41)
2021-06-21 00:19:13.668	[WARNING]  (1) : All workers exited. Exiting... (134)
2021-06-21 00:19:13.668	             rqf=44d08002 rqa=900 rpf=80000000 rpa=0 sif=EST,200020 sib=INI,30
2021-06-21 00:19:13.668	             ab=0,0 csb=0,0
2021-06-21 00:19:13.668	[ALERT]    (1) : exit-on-failure: killing every processes with SIGTERM
2021-06-21 00:19:13.667	             curr_task=0x7f68f349caa0 (task) calls=1 last=0
2021-06-21 00:19:13.667	             strm=0x7f68f33a6fa0 src=<SRC-IP> fe=ssl-unsecure-in be=bauth-be dst=unknown
2021-06-21 00:19:13.667	             af=0,0 csf=0x7f68f33d40f0,104000
2021-06-21 00:19:13.667	             stuck=1 prof=0 harmless=0 wantrdv=0
2021-06-21 00:19:13.667	             curr_task=0x7f68f36334a0 (task) calls=1 last=0
2021-06-21 00:19:13.667	             rqf=44d08002 rqa=900 rpf=80000000 rpa=0 sif=EST,200020 sib=INI,30
2021-06-21 00:19:13.667	             cob=0,0:NONE(0)/NONE(0)/NONE(0)
2021-06-21 00:19:13.667	 >Thread 2 : id=0x7f68f36d2b30 act=1 glob=0 wq=1 rq=1 tl=1 tlsz=2 rqsz=18
2021-06-21 00:19:13.667	             cpu_ns: poll=1219931253268 now=1221890237625 diff=1958984357
2021-06-21 00:19:13.667	Thread 1 is about to kill the process.
2021-06-21 00:19:13.667	             stuck=1 prof=0 harmless=0 wantrdv=0
2021-06-21 00:19:13.667	             cof=0x7f68f34cba60,a0003300:H1(0x7f68f34e86b0)/SSL(0x7f68f3588010)/tcpv4(22)
2021-06-21 00:19:13.667	*>Thread 1 : id=0x7f68f3f227a0 act=1 glob=1 wq=1 rq=1 tl=1 tlsz=1 rqsz=15
2021-06-21 00:19:13.667	             cpu_ns: poll=1188556814492 now=1191199005683 diff=2642191191
2021-06-21 00:19:13.667	               fct=0x55d1532a6200(process_stream) ctx=0x7f68f33a6fa0
2021-06-21 00:19:13.667	             ab=0,0 csb=0,0
```

**Comment by TimWolla** [comment]
>   `user bauth password <HASHED-PWD>`

What type of hash are you using? Please provide the prefix (first 10 characters or so). I'm suspecting that your hash is so expensive that it triggers the watch dog.

**Comment by lukastribus** [comment]
Yes, I think so too.

Haproxy needs to compute the hash for every single HTTP transaction, it's a bad idea to do this in the hot path of a load-balancer.

Modern hashes can be *very* expensive, because that is exactly their design goal.

**Comment by bsundmark** [comment]
If I remeber correctly it's a SHA-256: "$6$rounds=656000$3..."

It sounds very likely that this could be the problem. Would any of the other supported algorithms be better or should we try to implement it with insecure password. I think we could do that with our setup so that it's still secure enough.

Thanks for quick answer.

**Comment by TimWolla** [comment]
> Would any of the other supported algorithms be better or should we try to implement it with insecure password.

You could just try reducing the round of rounds to a lower value, e.g. `1000`. Depending on the amount of traffic you might need to go for insecure password. This should be sufficiently safe / acceptable if you use strong randomly generated passwords instead of some user input.

**Comment by wtarreau** [comment]
Wow I've never seen *that* many rounds on a hash! Indeed, it's a CPU killer! I would never have dug into that direction! Good catch, guys!

**Comment by bsundmark** [comment]
I use a small python script to generate the hash.
```
#!/usr/bin/env python3

# based on https://stackoverflow.com/a/17992126/117471

import sys
from getpass import getpass
from passlib.hash import sha512_crypt

passwd = input() if not sys.stdin.isatty() else getpass()
print(sha512_crypt.hash(passwd))
```
The rounds value 656000 is apparently the default for the sha_512 function used.

I updated to:
```
print(sha512_crypt.hash(passwd,rounds=1000))
```
1000 rounds is the minimum for sha_512 and hopefully it will work better. Will know in a few days when the batch job has run some more times.

Will update with result.
Thanks again for the help!

**Comment by TimWolla** [comment]
Thanks. I'll mark this issue as “invalid” and close it at this point, as it appears to be clear that this is not a bug in HAProxy. Looking forward to the results of your change, though!

**Comment by bsundmark** [comment]
Agree that it's probably not a bug.

Perhaps it might be a good idea to mention something in the documentation about this potential problem with basic auth.

**Comment by bsundmark** [comment]
Just wanted to report that we've had no issues since decreasing rounds on the pwd hash.

Case closed! 👍

**Comment by TimWolla** [comment]
> Just wanted to report that we've had no issues since decreasing rounds on the pwd hash.

Thank you for reporting back.

> Perhaps it might be a good idea to mention something in the documentation about this potential problem with basic auth.

I just checked the configuration doc, because it certainly should be mentioned there. However it already is: https://cbonte.github.io/haproxy-dconv/2.4/configuration.html#3.4-user

> Attention: Be aware that using encrypted passwords might cause significantly
> increased CPU usage, depending on the number of requests, and the algorithm
> used. For any of the hashed variants, the password for each request must
> be processed through the chosen algorithm, before it can be compared to the
> value specified in the config file. Most current algorithms are deliberately
> designed to be expensive to compute to achieve resistance against brute
> force attacks. They do not simply salt/hash the clear text password once,
> but thousands of times. This can quickly become a major factor in HAProxy's
> overall CPU consumption!

@bsundmark Do you feel this needs to be clarified better or did you just miss this paragraph?

**Comment by chipitsine** [comment]
using high rate basic auth seem to be an idea for testing corner cases.

I'm against using high rate in production because it is CPU waste and nothing more. but it can help to find some hidden bugs.

**Comment by bsundmark** [comment]
@TimWolla I think I saw that when we set it up but since our CPU utilization never exceeds a few percent and we experienced no problems for several months it didn't occur to us that the error we got was related to that. The error by the way didn't really help in understanding what the root cause was. In hindsight maybe we should have been able to put two and two together.

To improve things maybe add something to the warning that it could cause a watchdog reset and crash and how that manifests itself by the rather cryptic stack trace with words like "failure" and "Aborted".

@chipitsine As for using basic auth we had to in this case as the other party only supported that. We normally use certificate based auth.
